src/Features/Core/Portable/Completion/Providers/AbstractTypeImportCompletionProvider.cs

```diff
@@ -0,0 +1,322 @@
+using Microsoft.CodeAnalysis.Experiments;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.Internal.Log;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.Extensions.ContextQuery;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Completion.Providers
+{
+    internal abstract partial class AbstractTypeImportCompletionProvider : CommonCompletionProvider
+    {
+        private readonly SyntaxAnnotation _annotation = new SyntaxAnnotation();
+
+        protected abstract Task<SyntaxContext> CreateContextAsync(Document document, int position, CancellationToken cancellationToken);
+        protected abstract HashSet<INamespaceSymbol> GetNamespacesInScope(SyntaxNode location, SemanticModel semanticModel, CancellationToken cancellationToken);
+
+        private readonly ITypeImportCompletionService _typeImportCompletionService;
+        private readonly IExperimentationService _experimentationService;
+
+        public AbstractTypeImportCompletionProvider(Workspace workspace)
+            : base()
+        {
+            _typeImportCompletionService = workspace.Services.GetService<ITypeImportCompletionService>();
+            _experimentationService = workspace.Services.GetService<IExperimentationService>();
+        }
+
+        public override async Task ProvideCompletionsAsync(CompletionContext completionContext)
+        {
+#if DEBUG
+            DebugObject.DebugClear();
+            var tick = Environment.TickCount;
+#endif
+            var document = completionContext.Document;
+            var position = completionContext.Position;
+            var cancellationToken = completionContext.CancellationToken;
+
+            var importCompletionOptionValue = completionContext.Options.GetOption(CompletionOptions.ShowImportCompletionItems, document.Project.Language);
+
+            if (importCompletionOptionValue == false ||
+                importCompletionOptionValue == null && _experimentationService?.IsExperimentEnabled(WellKnownExperimentNames.TypeImportCompletion) != true)
+            {
+                return;
+            }
+
+            var syntaxContext = await CreateContextAsync(document, position, cancellationToken).ConfigureAwait(false);
+
+            var items = await GetCompletionItemsAsync(document, syntaxContext, position, cancellationToken).ConfigureAwait(false);
+            completionContext.AddItems(items);
+#if DEBUG
+            DebugObject.debug_total_time_with_ItemCreation = Environment.TickCount - tick;
+#endif
+            return;
+        }
+
+        private async Task<ImmutableArray<CompletionItem>> GetCompletionItemsAsync(Document document, SyntaxContext context, int position, CancellationToken cancellationToken)
+        {
+            if (context.IsTypeContext)
+            {
+                var project = document.Project;
+                var node = context.LeftToken.Parent;
```

Comment:
any chance 'node' could be null?

Reply:
You mean node or Left Token? In what case a token can have a null parent?

Reply:
leftToken could not exist, in which case Parent could be null.  probably should check for that